Kelpie stream10 dev bjs2 77219

JegorKuz opened the pull request:
Новая ветка без master в названии, с исправленным оформлением и логикой.

JegorKuz committed:
[UPDATE]
Решение для таски BJS2-77219.
Нет тестов.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Исправил ошибки по оформлению и логике.
Переделал фильтр в стратегию.
Заменил кастомный SQL "create" запрос на существующий "save".
Заменил реализации сервисом на интерфейсы.
Изменил выброс пустой ошибки, на ошибку с сообщением.

Bobritta commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+        Optional<RecommendationRequest> optionalRecommendationRequest = recommendationRequestRepository.findLatestPendingRequest(
+                recommendationRequestDto.getRequesterId(),
+                recommendationRequestDto.getReceiverId());
+
```
Если запрос не найден (то есть, если они никогда не делали запросов на рекомендации), будет выброшено исключение NoSuchElementException с сообщением о том, что либо запрашивающий, либо получатель отсутствуют в базе данных. Это произойдет, даже если сами пользователи существуют в базе данных, но у них нет записей о запросах на рекомендации, так?

JegorKuz replied:
Да, всё верно, запрос будет искать рекомендацию по пользователям, в 4ех случаях вернёт null, если:
1) Оба ид не найдены или отсутствуют.
2) 1 из 2ух ид отсутствует. (тут сразу 2 случая)
3) Оба присутствуют, но запроса ещё не было сделано.

Bobritta replied:
и тогда получится, что твой метод create вместо того, чтобы создать запрос, какого раньше не делал этот отправитель этому получателю, выбросит "Either requester or receiver are not present in data bases. Наверное, так не надо?

Bobritta commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+
+        if (recommendationRequest.getUpdatedAt().isBefore(LocalDateTime.now().minus(Period.ofMonths(6)))) {
+            boolean allSkillsExist = skillRequestService.skillRequestsExist(recommendationRequest.getSkills());
+
```
на задание "Также в методе нужно проверить, что запрашиваемые скиллы существуют в базе данных." я бы ожидала запрос к skillRepository, а не к skillRequestsRepository, пожалуй. Т.е. нам же всё равно запросы на какие навыки были созданы ранее, нам же важно какие навыки в принципе были созданы, да?

Bobritta replied:
Я вообще не пойму, откуда skillRequestsRepository взялся. В оригинальном наборе репозиториев его нет, а в изменённых файлах этого пул- рквеста - тоже. Но вижу, что ты его импортируешь в SkillRequestServiceImpl . Подскажи, куда нужно смотреть?

JegorKuz replied:
Первый комент:
Да, ты вроде верно всё подметила до точки.
После точки, где (Т.е ...), если ты про то что нам, надо посмотреть какие скилы в целом присутствуют в БД, то да. Тут проверка на то, что все те скилы которые есть в рекомендации есть и в БД обычных скиллов.

JegorKuz replied:
Касательно второго сообщения, путь: school.faang.user_service.repository.recommendation там есть SkillRequestRepository

Bobritta commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+            if (allSkillsExist) {
+                recommendationRequestRepository.save(recommendationRequest);
+
+                // Зачем повторно создавать скилы в базе данных если до этого мы проверяли их наличие в бд???
```
Вот потому что там была база всех существующих скилов, где мы проверяли, можно ли по ним делать запрос. А тут должна быть база скилов, по которым кто-нибудь хоть раз сделал запрос. Я так понимаю

JegorKuz replied:
Не совсем, тут идёт запрос в БД с SkillRequest, а просят их добавить в БД по Skill.

Bobritta replied:
Почему ты так думаешь? Там действительно в задаче нет, по крайней мере в этом пункте, уточнения, в какой базе мы должны проверять, но я бы рассуждала так:
есть 2 базы: Skill и SkillRequest,
в Skill мы кладём навыки, которые создают пользователи, а в SkillRequest - те, которые они запршивают в рекомендациях. Но для запроса рекомендации надо убедиться, что этот навык был когда-то кем-то создан, т.е. лежит в базе Skill.
Почему нет?

Bobritta reviewed (changes requested):
Проверила только первую часть "Запросить рекомендацию"

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Избавился от Optional изменив логику.
Дополнил решение для метода create - созданием скилов для entity.
Убрал не нужный SkillRequestService/Impl.
Добавил более подробные сообщения для Exceptions.
Убрал не нужные проверки в Controller.
Убрал не нужный метод из RecommendationRequestRepository.

[TODO]
Исправить фильтр.
До делать тесты.

Bobritta commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+                recommendationRequestDto.getSkillIds()
+                        .stream()
+                        .map(skillRepository::findById)
+                        .filter(Optional::isPresent)
```
ты же уже выше проверил, что они все существуют? Ты здесь какой-то особый случай ещё проверяешь?

JegorKuz replied:
Исправил, Михаил подметил что надо вернуть, вернул обратно)

Bobritta commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestService.java:
```diff
@@ -0,0 +1,18 @@
+
+    List<RecommendationRequestDto> getRequests(RequestFilterDto filter);
+
+    RecommendationRequestDto getRequest(long id);
```
Миша и вообще все решили, что Long везде в ДТО должен заменить long, даже если примитив стоял по условию задачию. И у тебя в ДТО вроде обертка, а тут почему-то - нет

miked4king replied:
Все правильно, у нас запрос инфромации по id, скорее всего этот id будет передаваться в пути запроса, в таком случае мы принимаем long

Bobritta commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+    public RecommendationRequestDto getRequest(long id) {
+        return recommendationRequestMapper.toDto(recommendationRequestRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException("Recommendation request with id %s doesn't exist".formatted(id))));
+    }
```
Миша рекомендовал использовать EntityNotFoundException вместо IllegalArgument, по идее у тебя тут тоже энтити

Bobritta replied:
3-я часть "Получить конкретный запрос на рекомендацию"
5.	В методе getRequest(RequestFilterDto filter) класса RecommendationRequestService нужно получить нужную рекомендацию из базы, используя метод findById класса RecommendationRequestRepository. Если такая рекомендация не найдена, выбросить исключение.

Тут, когда сделаешь фильтры, можно будет реализовать эту логику. Хотя мне кажется, что текущий вариант оптимальнее. Нет смысла грузить всю базу, затем пропускать через фильтр поиска id, если у репозитория уже есть удобный метод запроса к базе. Но, поскольку, в условии прам оговорено, что надо в фильтр поместить, лучше у Миши, наверное, уточнить

Bobritta commented on src/main/java/school/faang/user_service/filter/recommendation/RecommendationRequestFilterBuilder.java:
```diff
@@ -0,0 +1,36 @@
+
+    public static List<RecommendationRequestFilterStrategy> buildStrategies(RequestFilterDto filter) {
+        List<RecommendationRequestFilterStrategy> strategies = new ArrayList<>();
+
```
https://faang-school.com/courses/4jnzmndg/9nzw6rrd
Посмотри воркшоп из 5.2 по использованию фильтров. Это и правда больше по СОЛИД, и разжёвано всё прям. Думаю, всё равно будут именно такого требовать и Мише в любом случае придётся тебе вернуть такой код

JegorKuz replied:
Исправил

Bobritta reviewed (changes requested):
Проверила вторую часть. С фильтрами надо переделывать, я думаю. Создавать лист фильтров, каждое поле проверять отдельно, чтобы избежать каскада ифов. В видео из 5.2 есть решение

Bobritta commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+        recommendationRequest.setStatus(RequestStatus.REJECTED);
+        recommendationRequest.setRejectionReason(rejection.getReason());
+
+        return recommendationRequestMapper.toDto(recommendationRequest);
```
1. в RecommendationRequestDto, который ты возвращаешь нет поля RejectionReason. А значит получатель не увидит причину.  Может, и возвращать надо RejectionDto, содержащий id исходного запроса и причину отказа?

2. М.б. RejectionDto стоит содержать id исходного запроса и причину отказа ещё и вот по какому соображению:
в моей задаче и других тоже, когда id передаётся отдельным аргументом от ДТО, это означает, что его мы извлекаем из адреса запрашивающего, чтобы проверить, что он- именно тот человек, которому позволено отклонить запрос (или, в моей задаче, обновить именно своё расписание).
Так что плюсом к этому пункту, возможно стоит проверить, что этот id (в таком случае он относится к юзеру, а не к запросу), соответствует receiverId запроса, чей id содержался бы как поле внутри RejectionDto  вместе с причиной отказа

Bobritta replied:
...  вернуть надо причину. Ты её в энтити запроса-то вставляешь, но в возвращаемом ДТО - нет

JegorKuz replied:
Обсудили на созвоне, получается на причину пофиг при возвращении дто.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Исправио фильтр.

[TODO]
До делать тесты.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Убрал не нужный чек для create.

Bobritta reviewed (approved):
для уточнения деталей задания у техлида

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+    private final SkillRequestRepository skillRequestRepository;
+    private final RecommendationRequestRepository recommendationRequestRepository;
+    private final List<RecommendationRequestFilterStrategy> recommendationRequestFilters;
+
```
лишний отступ

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+
+    private final RecommendationRequestMapper recommendationRequestMapper;
+
+    public RecommendationRequestDto create(RecommendationRequestDto recommendationRequestDto) {
```
при реализации методов интерфейса принято добавлять над методами аннотацию @Override, давай добавим.
в методах ниже тоже

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+        User requester = userRepository.findById(recommendationRequestDto.getRequesterId())
+                .orElseThrow(() -> new IllegalArgumentException("Requester with id %s was not found".formatted(recommendationRequestDto.getRequesterId())));
+        User receiver = userRepository.findById(recommendationRequestDto.getReceiverId())
+                .orElseThrow(() -> new IllegalArgumentException("Receiver with id %s was not found".formatted(recommendationRequestDto.getReceiverId())));
```
лучше подойдет EntityNotFoundException, т.к в базе сущность найти не можем
в других местах, где не получается найти сущность в репозитории давай тоже поправим

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+        recommendationRequest.setRequester(requester);
+        recommendationRequest.setReceiver(receiver);
+
+        if (!recommendationRequest.getUpdatedAt().isBefore(LocalDateTime.now().minus(Period.ofMonths(6)))) {
```
может упростим условие убрав отрицание "!" и используем isAfter ?

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+                        .map(optionalSkill ->
+                                skillRequestRepository.create(
+                                        recommendationRequest.getId(),
+                                        optionalSkill.get().getId()))
```
понимаю что проверял что все скилы есть в 54 строке.
но перед получением значения из Optional переменной всегда требуется проверка, что Optional не пустой.
давай добавим

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        return recommendations
+                .map(recommendationRequestMapper::toDto)
```
map можно не выносить на новую строку

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+            throw new IllegalArgumentException("Recommendation request has already been updated in the last 6 months.");
+        }
+
+        if (!recommendationRequestDto.getSkillIds().stream().allMatch(skillRepository::existsById)) {
```
не очень практика передавать стримы в условие if, давай лучше сохраним результат стрима в переменную с говорящим названием и передадим ее в if

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+            throw new IllegalArgumentException("Unable to reject request");
+        }
+
+        // тут не сохраняю в БД значит метод не меняет ни как логику, можно обновить страницу и всё будет по старому
```
а почему не сохраняем в бд? какой смысл тогда от данного метода?
получается мы вернем пользователю ответ что запрос отклонен, а по факту нет)

JegorKuz replied:
Изменил, теперь сохраняем

miked4king commented on src/main/java/school/faang/user_service/filter/recommendation/filters/StatusFilter.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    public Stream<RecommendationRequest> apply(Stream<RecommendationRequest> recommendationRequests, RequestFilterDto requestFilterDto) {
+        return recommendationRequests
+                .filter(recommendationRequest ->
```
filter можно не переносить на новую строку

miked4king commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+        return recommendationRequestMapper.toDto(recommendationRequestRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException("Recommendation request with id %s doesn't exist".formatted(id))));
+    }
+
```
лишний отступ

miked4king commented:
В целом хорошая работа, названия, форматирование, обработка ошибок, красавчик!
оставил немного комментариев

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Исправил ошибки по комментариям.
